examples/ethereum-tracker/src/contract.rs

```diff
@@ -46,10 +46,16 @@ impl Contract for EthereumTrackerContract {
             contract_address,
             start_block,
         } = argument;
+
         self.state.ethereum_endpoint.set(ethereum_endpoint);
         self.state.contract_address.set(contract_address);
         self.state.start_block.set(start_block);
-        self.read_initial(start_block).await;
+        self.state
+            .save()
+            .await
+            .expect("Failed to write updated storage");
```

Comment:
What is exactly the justification for adding this `self.state.save()` here? This is the only smart contract where a `self.state.save()` is added to the `fn instantiate` function. Why is that so?
For example, the `crowd-funding` and `fungible` are doing meaningful operations in their `fn instantiate` functions. Why haven't they some `self.state.save()`.
Side question 1: When is the `fn store` function used?
Side question 2: Now, the `fn save` function is doing the writing to the batch and if the batch is empty, then nothing is done. So, could the `save`  could be added by the linera-rpc?

Reply:
Why doesn't service access the cache (which is an overlay over persistent state + the things that contract did) ?

Reply:
Because the cache is in memory inside the contract's Wasm environment, and when the service is called a separate Wasm environment is created. The way the update is sent from the contract to the service is by "persisting" data (which actually just updates the cache in the host's memory, which is only truly persisted if the block executes successfully).

Reply:
No, A will only see the response B sent to it. If A calls B again, then B's changes will still be available because its runtime is reused. This is also why if A calls B and then B calls A, A doesn't have to save its state before calling B.

I don't think persisting before queries would be a good idea. The problem is that if we do that invisibly, we might cause bugs where unexpected state is persisted. It's a tricky situation :confused: I'm starting to think splitting the runtime function into two functions with more verbose names, like `save_state_and_query_service` and `query_service_without_saving_state`, would at least reduce confusion and the risk of (potentially serious) state persistence bugs :thinking: